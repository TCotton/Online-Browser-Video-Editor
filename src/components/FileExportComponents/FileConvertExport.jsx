import React, {useState} from 'react';
import {createFFmpeg, fetchFile} from "@ffmpeg/ffmpeg";
import {useLiveQuery} from "dexie-react-hooks";
import {document, window} from "browser-monads";
import db from "../indexDB/indexDB";
import dbVF from "../indexDB/indexDBVideo";
import {changeExt} from "../helperFunctions";

function downloadBlob(blobUrl, name) {
    // Create a link element
    const link = document.createElement("a");

    // Set link's href to point to the Blob URL
    link.href = blobUrl;
    link.download = name;

    // Append link to the body
    document.body.appendChild(link);

    // Dispatch click event on the link
    // This is necessary as link.click() does not work on the latest firefox
    link.dispatchEvent(
        new MouseEvent('click', {
            bubbles: true,
            cancelable: true,
            view: window
        })
    );

    // Remove link from body
    document.body.removeChild(link);
}

export const FileConvertExport = () => {
    const [status, setStatus] = useState('');
    const allFileItems = useLiveQuery(() => db.file.toArray(), []);
    const allVideoItems = useLiveQuery(() => dbVF.videofile.toArray(), []);

    const ffmpeg = createFFmpeg({
        log: true, progress: ({ratio}) => {
            if ((ratio * 100.0).toFixed(2) < 99.9) {
                setStatus(`Complete: ${(ratio * 100.0).toFixed(2)}%`);
            }
        }
    });

    const onClickTwitter = async () => {
        const result = window.confirm('Convert video for Twitter?');
        if (!result) return;

        console.info('Loading ffmpeg-core.js');

        await ffmpeg.load();
        setStatus('Video transcoding started');

        ffmpeg.FS('writeFile', allVideoItems[0].name, await fetchFile(allFileItems[0]));

        await ffmpeg.run('-i',
            allVideoItems[0].name,
            "-vcodec",
            "libx264",
            "-vf",
            "scale=640:-1",
            "-pix_fmt",
            "yuv420p",
            "-strict",
            "experimental",
            "-r",
            "30",
            "-t",
            "2:20",
            "-acodec",
            "aac",
            "-vb",
            "1024k",
            "-minrate",
            "1024k",
            "-maxrate",
            "1024k",
            "-bufsize",
            "1024k",
            "-ar",
            "44100",
            "-ac",
            "2",
            `converted-${allVideoItems[0].name}`);


        setStatus('Video transcoding ended');
        const data = ffmpeg.FS('readFile', `converted-${allVideoItems[0].name}`);
        const href = window.URL.createObjectURL(new Blob([data.buffer], {type: 'video/mp4'}));
        downloadBlob(href, changeExt(allVideoItems[0].name, 'mp4'));
        window.URL.revokeObjectURL(href);
    }

    const onClickFacebook = async () => {
        const result = window.confirm('Convert video for Facebook?');
        if (!result) return;

        console.info('Loading ffmpeg-core.js');
        await ffmpeg.load();
        setStatus('Video transcoding started');
        ffmpeg.FS('writeFile', allVideoItems[0].name, await fetchFile(allFileItems[0]));
        await ffmpeg.run('-i',
            allVideoItems[0].name,
            "-c:v",
            "libx264",
            "-acodec",
            "aac",
            "-crf",
            "18",
            "-preset",
            "faster",
            "-c:a",
            "copy",
            `converted-${allVideoItems[0].name}`);
        setStatus('Video transcoding ended');
        const data = ffmpeg.FS('readFile', `converted-${allVideoItems[0].name}`);
        const href = window.URL.createObjectURL(new Blob([data.buffer], {type: 'video/mp4'}));
        downloadBlob(href, changeExt(allVideoItems[0].name, 'mp4'));
        window.URL.revokeObjectURL(href);
    }

    const onClickLinkedin = async () => {
        const result = window.confirm('Convert to video for Linkedin?');
        if (!result) return;

        console.info('Loading ffmpeg-core.js');
        await ffmpeg.load();
        setStatus('Video transcoding started');
        ffmpeg.FS('writeFile', allVideoItems[0].name, await fetchFile(allFileItems[0]));
        await ffmpeg.run('-i',
            allVideoItems[0].name,
            "-c:v",
            "libx264",
            "-acodec",
            "aac",
            "-crf",
            "18",
            "-preset",
            "faster",
            "-c:a",
            "copy",
            `converted-${allVideoItems[0].name}`);
        setStatus('Video transcoding ended');
        const data = ffmpeg.FS('readFile', `converted-${allVideoItems[0].name}`);
        const href = window.URL.createObjectURL(new Blob([data.buffer], {type: 'video/mp4'}));
        downloadBlob(href, changeExt(allVideoItems[0].name, 'mp4'));
        window.URL.revokeObjectURL(href);
    }

    return (
        <>
            <div className="export" data-testid="export">
                <header>
                    <p className="title">Click below to export</p>
                    <p className="status">{status}</p>
                </header>
                <ul className="social">
                    <li>
                        <button onClick={onClickTwitter} data-testid="twitter">
                            <img src={"./social/twitter.svg"} alt="click to export for twitter" width="400"
                                 height="400" />
                        </button>
                    </li>
                    <li>
                        <button onClick={onClickFacebook} data-testid="facebook">
                            <img src={"./social/facebook.png"} alt="click to export for Facebook" width="250"
                                 height="250"/>
                        </button>
                    </li>
                    <li>
                        <button onClick={onClickLinkedin} data-testid="linkedin">
                            <img src={"./social/linkedin.png"} alt="click to export for Linkedin" width="650"
                                 height="540"/>
                        </button>
                    </li>
                </ul>
            </div>
        </>
    )
}