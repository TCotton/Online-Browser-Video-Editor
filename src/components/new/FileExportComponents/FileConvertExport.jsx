import React, {useState} from 'react';
import {createFFmpeg, fetchFile} from "@ffmpeg/ffmpeg";
import {useLiveQuery} from "dexie-react-hooks";
import {document, window} from "browser-monads";
import db from "../indexDB/indexDB";
import dbVF from "../indexDB/indexDBVideo";

function downloadBlob(blobUrl, name) {
    // Create a link element
    const link = document.createElement("a");

    // Set link's href to point to the Blob URL
    link.href = blobUrl;
    link.download = name;

    // Append link to the body
    document.body.appendChild(link);

    // Dispatch click event on the link
    // This is necessary as link.click() does not work on the latest firefox
    link.dispatchEvent(
        new MouseEvent('click', {
            bubbles: true,
            cancelable: true,
            view: window
        })
    );

    // Remove link from body
    document.body.removeChild(link);
}

export const FileConvertExport = () => {
    const [status, setStatus] = useState('');
    const allFileItems = useLiveQuery(() => db.file.toArray(), []);
    const allVideoItems = useLiveQuery(() => dbVF.videofile.toArray(), []);

    const onClickTwitter = async () => {
        const result = window.confirm('Convert video for Twitter?');
        if (!result) return;

        const ffmpeg = createFFmpeg({
            log: true, progress: ({ratio}) => {
                if ((ratio * 100.0).toFixed(2) < 99.9) {
                    setStatus(`Complete: ${(ratio * 100.0).toFixed(2)}%`);
                }

            }
        });

        console.info('Loading ffmpeg-core.js');
        await ffmpeg.load();
        setStatus('Video transcoding started');
        ffmpeg.FS('writeFile', allVideoItems[0].name, await fetchFile(allFileItems[0]));
        await ffmpeg.run('-i',
            allVideoItems[0].name,
            "-vcodec",
            "libx264",
            "-vf",
            "scale=640:-1",
            "-pix_fmt",
            "yuv420p",
            "-strict",
            "experimental",
            "-r",
            "30",
            "-t",
            "2:20",
            "-acodec",
            "aac",
            "-vb",
            "1024k",
            "-minrate",
            "1024k",
            "-maxrate",
            "1024k",
            "-bufsize",
            "1024k",
            "-ar",
            "44100",
            "-ac",
            "2",
            `converted-${allVideoItems[0].name}`);
        setStatus('Video transcoding ended');
        const data = ffmpeg.FS('readFile', `converted-${allVideoItems[0].name}`);
        const href = window.URL.createObjectURL(new Blob([data.buffer], {type: 'video/mp4'}));
        downloadBlob(href, allVideoItems[0].name);
        window.URL.revokeObjectURL(href);
    }

    const onClickFacebook = async () => {
        const result = window.confirm('Convert video for Facebook?');
        if (!result) return;

        const ffmpeg = createFFmpeg({
            log: true, progress: ({ratio}) => {
                if ((ratio * 100.0).toFixed(2) < 99.9) {
                    setStatus(`Complete: ${(ratio * 100.0).toFixed(2)}%`);
                }

            }
        });

        console.info('Loading ffmpeg-core.js');
        await ffmpeg.load();
        setStatus('Video transcoding started');
        ffmpeg.FS('writeFile', allVideoItems[0].name, await fetchFile(allFileItems[0]));
        await ffmpeg.run('-i',
            allVideoItems[0].name,
            "-c:v",
            "libx264",
            "-crf",
            "18",
            "-preset",
            "fast",
            "-c:a",
            "copy",
            `converted-${allVideoItems[0].name}`);
        setStatus('Video transcoding ended');
        const data = ffmpeg.FS('readFile', `converted-${allVideoItems[0].name}`);
        const href = window.URL.createObjectURL(new Blob([data.buffer], {type: 'video/mp4'}));
        downloadBlob(href, allVideoItems[0].name);
        window.URL.revokeObjectURL(href);
    }

    const onClickLinkedin = async () => {
        const result = window.confirm('Convert to video for Linkedin?');
        if (!result) return;

        const ffmpeg = createFFmpeg({
            log: true, progress: ({ratio}) => {
                if ((ratio * 100.0).toFixed(2) < 99.9) {
                    setStatus(`Complete: ${(ratio * 100.0).toFixed(2)}%`);
                }

            }
        });

        console.info('Loading ffmpeg-core.js');
        await ffmpeg.load();
        setStatus('Video transcoding started');
        ffmpeg.FS('writeFile', allVideoItems[0].name, await fetchFile(allFileItems[0]));
        await ffmpeg.run('-i',
            allVideoItems[0].name,
            "-c:v",
            "libx264",
            "-crf",
            "18",
            "-preset",
            "fast",
            "-c:a",
            "copy",
            `converted-${allVideoItems[0].name}`);
        setStatus('Video transcoding ended');
        const data = ffmpeg.FS('readFile', `converted-${allVideoItems[0].name}`);
        const href = window.URL.createObjectURL(new Blob([data.buffer], {type: 'video/mp4'}));
        downloadBlob(href, allVideoItems[0].name);
        window.URL.revokeObjectURL(href);
    }

    return (
        <>
            <div className={`export`}>
                <p>Click below to export (optimises for those platforms)</p>
                <p>{status}</p>
                <ul className="social">
                    <li>
                        <button onClick={onClickTwitter}>
                            <img src={`./social/twitter.svg`} alt={`click to export for twitter`} width="400"
                                 height="400"/>
                        </button>
                    </li>
                    <li>
                        <button onClick={onClickFacebook}>
                            <img src={`./social/facebook.png`} alt={`click to export for Facebook`} width="250"
                                 height="250"/>
                        </button>
                    </li>
                    <li>
                        <button onClick={onClickLinkedin}> d
                            <img src={`./social/linkedin.png`} alt={`click to export for Linkedin`} width="650"
                                 height="540"/>
                        </button>
                    </li>
                </ul>
            </div>
        </>
    )
}