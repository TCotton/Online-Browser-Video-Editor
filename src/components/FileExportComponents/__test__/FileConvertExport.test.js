import React from 'react'
import { window } from 'browser-monads'
import { fireEvent, render, waitFor } from '@testing-library/react'
import { FileConvertExport } from '../FileConvertExport'

jest.mock('@ffmpeg/ffmpeg', () => {
  return {
    createFFmpeg: () => {
      return {
        setProgress: jest.fn(),
        setLogger: jest.fn(),
        setLogging: jest.fn(),
        load: () =>
          Promise.resolve(() => {
            return true
          }),
        isLoaded: jest.fn(),
        run: jest.fn(),
        FS: jest.fn(n => n)
      }
    },
    fetchFile: () =>
      Promise.resolve(() => {
        return true
      })
  }
})

jest.mock('dexie-react-hooks', () => ({
  ...jest.requireActual('dexie-react-hooks'),
  useLiveQuery: jest.fn(n => {
    return n
  })
}))

describe('', () => {
  const windowAlert = jest.fn()
  const consoleInfo = jest.fn()
  const windowConfirm = jest.fn()
  const mockCreatObjectURL = jest.fn()
  const mockRevokeObjectURL = jest.fn()

  beforeEach(() => {
    window.alert = windowAlert
    window.console.info = consoleInfo
    window.confirm = windowConfirm
    window.URL.createObjectURL = mockCreatObjectURL
    window.URL.revokeObjectURL = mockRevokeObjectURL
  })

  afterEach(() => {
    jest.resetAllMocks()
  })

  it('should render successfully', () => {
    const { container } = render(<FileConvertExport />)
    expect(container.querySelector('[data-testid="export"]')).not.toBeNull()
  })

  it('should match snapshot', () => {
    const { baseElement } = render(<FileConvertExport />)
    expect(baseElement).toMatchSnapshot()
  })

  //TODO: Come back to this test
  xit('should call window confirm when clicking on the twitter image', async () => {
    const { container } = render(<FileConvertExport />)
    const twitter = container.querySelector('[data-testid="twitter"]')
    window.confirm = jest.fn(() => true)

    await waitFor(() => {
      fireEvent.click(twitter)
    })

    expect(window.confirm).toHaveBeenCalledTimes(1)
    expect(window.console.info).toHaveBeenCalledTimes(1)
    expect(window.URL.createObjectURL).toHaveBeenCalledTimes(1)
    expect(window.URL.revokeObjectURL).toHaveBeenCalledTimes(1)
  })
})
