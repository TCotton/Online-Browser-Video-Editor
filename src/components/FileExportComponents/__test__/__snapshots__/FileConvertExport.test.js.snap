// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<body>
  <div>
    <div
      class="export"
      data-testid="export"
    >
      <header>
        <p
          class="title"
        >
          Click below to export
        </p>
        <p
          class="status"
        />
      </header>
      <ul
        class="social"
      >
        <li>
          <button
            data-testid="twitter"
          >
            <img
              alt="click to export for twitter"
              height="400"
              src="./social/twitter.svg"
              width="400"
            />
          </button>
        </li>
        <li>
          <button
            data-testid="facebook"
          >
            <img
              alt="click to export for Facebook"
              height="250"
              src="./social/facebook.png"
              width="250"
            />
          </button>
        </li>
        <li>
          <button
            data-testid="linkedin"
          >
            <img
              alt="click to export for Linkedin"
              height="540"
              src="./social/linkedin.png"
              width="650"
            />
          </button>
        </li>
      </ul>
    </div>
  </div>
</body>
`;
